### PR TITLE
swap Incorrect balance value of To token [sc-15663]

### PR DESCRIPTION
**Pull request by vitaliipustovyi**

### CODE REVIEW

File: apps/nfid-frontend/src/features/transfer-modal/components/receive.tsx [hunk 1]
```diff
@@ -26,38 +21,25 @@ export const TransferReceive = ({
...
+      AccountIdentifier.fromPrincipal({
+        principal: Principal.fromText(publicKey),
+      }).toHex(),
+    )
 
-      setIsAccountLoading(false)
-    }
-  }, [vaultsAccountsOptions, isVault, publicKey, isVaultsAccountsValidating])
+    setIsAccountLoading(false)
+    //}
```
Comment (olsemeno) [inline apps/nfid-frontend/src/features/transfer-modal/components/receive.tsx:33]: comments

File: apps/nfid-frontend/src/integration/ft/impl/ft-impl.ts [hunk 2]
```diff
@@ -42,6 +42,10 @@ export class FTImpl implements FT {
     return this
   }
 
+  isInited(): boolean {
+    return this.tokenBalance !== undefined
```
Comment (olsemeno) [inline apps/nfid-frontend/src/integration/ft/impl/ft-impl.ts:46]: let's add. boolean field `inited`

File: apps/nfid-frontend/src/integration/nft/impl/nft-abstract.ts [hunk 3]
```diff
@@ -43,6 +43,10 @@ export abstract class NftImpl implements NFT {
     return this
   }
 
+  isInited(): boolean {
+    return this.assetPreview !== undefined
```
Comment (olsemeno) [inline apps/nfid-frontend/src/integration/nft/impl/nft-abstract.ts:47]: same here

**Review by olsemeno — APPROVED** [review]

File: apps/nfid-frontend/src/integration/ft/impl/ft-impl.ts (hunk 2 above)
Comment (vitaliipustovyi) [reply apps/nfid-frontend/src/integration/ft/impl/ft-impl.ts:46]: it's already return boolean.
Didn't get you.
Did you mean return object { inited: boolean } ?
@olekid

Comment (olekid) [reply apps/nfid-frontend/src/integration/ft/impl/ft-impl.ts:46]: Oleksii means to change method isInited to boolean field inited on your class

File: packages/ui/src/molecules/choose-modal/ft-modal.tsx [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+    if (originalIndex !== -1 && !options[originalIndex].isInited()) {
+      const { publicKey } = await getUserPrincipalId()
+
+      const updatedToken = await options[originalIndex].init(
+        Principal.fromText(publicKey),
+      )
+
+      setTokensOptions((prev) => {
+        const newOptions = prev.map((token, i) =>
```
Comment (olekid) [inline packages/ui/src/molecules/choose-modal/ft-modal.tsx:92]: not critical, but map is bad usage here, its used to map array to change its values, in your case just copy array and change value by index

File: packages/ui/src/molecules/choose-modal/nft-modal.tsx [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+        nft
+          .getCollectionName()
+          .toLowerCase()
+          .includes(searchInput.toLowerCase()) ||
+        nft.getTokenName().toLowerCase().includes(searchInput.toLowerCase()),
+    )
+  }, [nftsOptions, searchInput])
+
+  useIntersectionObserver(itemRefs.current, async (index) => {
+    const originalIndex = options.findIndex(
```
Comment (olekid) [inline packages/ui/src/molecules/choose-modal/nft-modal.tsx:75]: how can originalIndex be different from index?

File: packages/ui/src/molecules/choose-modal/nft-modal.tsx [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    )
+
+    if (originalIndex !== -1 && !options[originalIndex].isInited()) {
+      const updatedNft = await options[originalIndex].init()
+
+      setNftsOptions((prev) => {
+        const newOptions = prev.map((nft, i) =>
+          i === originalIndex ? updatedNft : nft,
+        )
+        return [...new Set(newOptions)]
```
Comment (olekid) [inline packages/ui/src/molecules/choose-modal/nft-modal.tsx:86]: why Set is used here? and again map ca be removed

File: packages/ui/src/molecules/choose-modal/nft-modal.tsx [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+            "scrollbar-thumb-rounded-full scrollbar-track-rounded-full",
+          )}
+        >
+          {filteredNfts.map((nft, index) => (
+            <div
+              id={`option_${nft.getTokenName().replace(/\s/g, "")}`}
+              key={`group_${nft.getCollectionId()}_${nft.getTokenName()}_${index}`}
+              ref={(el) => (itemRefs.current[index] = el)}
+            >
+              <ChooseNftItem
```
Comment (olekid) [inline packages/ui/src/molecules/choose-modal/nft-modal.tsx:157]: if these two modals only differ in option component being rendered it definitely makes sense to not duplicate such a big piece of code, you could have one component for this purpose and pass ChooseNftItem as an optionComponent prop to modal

File: packages/ui/src/molecules/choose-modal/token-modal.tsx [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+              "scrollbar-thumb-rounded-full scrollbar-track-rounded-full",
+            )}
+          >
+            {filteredTokens.map((token, index) => (
+              <div
+                key={`group_${token.getTokenName()}_${index}`}
+                ref={(el) => (itemRefs.current[index] = el)}
+                onClick={() => handleSelect(token)}
+              >
+                {renderItem(token, index)}
```
Comment (olekid) [inline packages/ui/src/molecules/choose-modal/token-modal.tsx:160]: seems key is not needed in option components

File: packages/ui/src/molecules/choose-modal/ft-modal.tsx [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    <>
+      <ChooseTokenModal
+        tokens={tokens}
+        title={title}
+        filter={filterTokens}
+        onSelect={onSelect}
+        trigger={trigger}
+        renderItem={(token, index) => (
+          <ChooseFtItem
+            key={`option_${token.getTokenAddress()}_group_${index}`}
```
Comment (olekid) [inline packages/ui/src/molecules/choose-modal/ft-modal.tsx:41]: key is not needed

File: packages/ui/src/molecules/choose-modal/ft-modal.tsx [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+  return (
+    <>
+      <ChooseTokenModal
+        tokens={tokens}
+        title={title}
+        filter={filterTokens}
+        onSelect={onSelect}
+        trigger={trigger}
+        renderItem={(token, index) => (
```
Comment (olekid) [inline packages/ui/src/molecules/choose-modal/ft-modal.tsx:39]: change function to component that receives prop token

File: packages/ui/src/molecules/choose-modal/ft-modal.tsx [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+  trigger?: JSX.Element
+}
+
+export const ChooseFtModal = ({
+  tokens,
+  onSelect,
+  title,
+  trigger,
+}: IChooseFtModal) => {
+  const filterTokens = useCallback((token: FT, searchInput: string) => {
```
Comment (olekid) [inline packages/ui/src/molecules/choose-modal/ft-modal.tsx:21]: filterTokensBySearchInput

File: packages/ui/src/molecules/choose-modal/ft-modal.tsx [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+      token.getTokenName().toLowerCase().includes(searchInput.toLowerCase())
+    )
+  }, [])
+
+  return (
+    <>
+      <ChooseTokenModal
+        tokens={tokens}
+        title={title}
+        filter={filterTokens}
```
Comment (olekid) [inline packages/ui/src/molecules/choose-modal/ft-modal.tsx:36]: filterTokensBySearchInput

File: packages/ui/src/molecules/choose-modal/nft-modal.tsx [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+    <>
+      <ChooseTokenModal
+        tokens={tokens}
+        title={title}
+        filter={filterTokens}
+        onSelect={onSelect}
+        trigger={trigger}
+        renderItem={(token, index) => (
+          <ChooseNftItem
+            key={`option_${token.getTokenId()}_group_${index}_${index}`}
```
Comment (olekid) [inline packages/ui/src/molecules/choose-modal/nft-modal.tsx:41]: key

File: packages/ui/src/molecules/choose-modal/token-modal.tsx [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+        if (tokenIndex !== -1) {
+          newOptions[tokenIndex] = updatedToken as T
+        }
+
+        return newOptions
+      })
+    }
+  })
+
+  const handleSelect = useCallback(
```
Comment (olekid) [inline packages/ui/src/molecules/choose-modal/token-modal.tsx:94]: NftModal and FtModal can handle this case on their own

**Review by olekid — APPROVED** [review]

**Pull request merged**